File: apps/react-native/src/services/native-passkey-auth.ts
import { Buffer } from 'buffer';
import { keccak256, sha256, toBytes } from 'viem';

import {
  litRustNativeCreatePasskey,
  litRustNativeGetPasskey,
  litRustTestConnect,
} from './LitRustNative';

const AUTH_METHOD_TYPE_WEBAUTHN = 3;
const DEFAULT_SCOPES = ['sign-anything', 'personal-sign'];
const BLOCKHASH_URL = 'https://block-indexer.litgateway.com/get_most_recent_valid_block';
const MINT_STATUS_MAX_ATTEMPTS = 20;
const MINT_STATUS_POLL_DELAY_MS = 3000;

export type NativePasskeyAuthData = {
  authMethodType: number;
  authMethodId: string;
  accessToken: string;
};

export type NativePasskeyRegisterAndMintResult = {
  pkpInfo: unknown;
  webAuthnPublicKey: string;
  authData: NativePasskeyAuthData;
};

function sleep(ms: number): Promise<void> {
  return new Promise((resolve) => setTimeout(resolve, ms));
}

function requireString(value: unknown, fieldName: string): string {
  const parsed = typeof value === 'string' ? value.trim() : '';
  if (!parsed) throw new Error(`${fieldName} cannot be empty`);
  return parsed;
}

function sanitizeBaseUrl(value: string): string {
  return requireString(value, 'authServiceBaseUrl').replace(/\/+$/, '');
}

function buildAuthServiceHeaders(passkeyRpId: string): Record<string, string> {
  const origin = `https://${passkeyRpId}`;
  return {
    Origin: origin,
    Referer: `${origin}/`,
  };
}

function parseHostname(value: string, fieldName: string): string {
  const trimmed = requireString(value, fieldName);
  const withScheme = /^[a-z][a-z0-9+.-]*:\/\//i.test(trimmed) ? trimmed : `https://${trimmed}`;
  let parsed: URL;
  try {
    parsed = new URL(withScheme);
  } catch {
    throw new Error(`Invalid ${fieldName}: ${value}`);
  }
  const hostname = parsed.hostname.trim().toLowerCase();
  if (!hostname) throw new Error(`Invalid ${fieldName}: ${value}`);
  return hostname;
}

function normalizeRpIdHost(value: string): string {
  return parseHostname(value, 'passkeyRpId');
}

function hexToBytes(value: string): Uint8Array {
  const normalized = value.startsWith('0x') ? value.slice(2) : value;
  if (!/^[0-9a-fA-F]+$/.test(normalized) || normalized.length % 2 !== 0) {
    throw new Error(`Invalid hex value: ${value}`);
  }
  const bytes = new Uint8Array(normalized.length / 2);
  for (let index = 0; index < normalized.length; index += 2) {
    bytes[index / 2] = parseInt(normalized.slice(index, index + 2), 16);
  }
  return bytes;
}

function toBase64Url(value: Uint8Array): string {
  const base64 = Buffer.from(value).toString('base64');
  return base64.replace(/\+/g, '-').replace(/\//g, '_').replace(/=+$/g, '');
}

function base64UrlToBytes(value: string): Uint8Array {
  const normalized = value.replace(/-/g, '+').replace(/_/g, '/');
  const padded = normalized + '='.repeat((4 - (normalized.length % 4)) % 4);
  return new Uint8Array(Buffer.from(padded, 'base64'));
}

function canonicalizeBase64Like(value: string): string {
  return toBase64Url(base64UrlToBytes(value));
}

function canonicalizeCredentialId(value: string): string {
  return canonicalizeBase64Like(value);
}

function bytesToHex(value: Uint8Array): string {
  return Buffer.from(value).toString('hex');
}

function deriveWebAuthnAuthMethodId(rawId: string): string {
  return keccak256(toBytes(`${rawId}:lit`));
}

function inspectWebAuthnAssertion(
  authenticationResponse: Record<string, unknown>,
  expectedRpId: string,
): {
  expectedRpIdHashHex: string;
  authenticatorRpIdHashHex?: string;
  rpIdHashMatchesExpected: boolean;
  clientDataOrigin?: string;
  clientDataOriginHost?: string;
} {
  const expectedRpIdHashHex = sha256(toBytes(expectedRpId)).slice(2).toLowerCase();
  const inspection: {
    expectedRpIdHashHex: string;
    authenticatorRpIdHashHex?: string;
    rpIdHashMatchesExpected: boolean;
    clientDataOrigin?: string;
    clientDataOriginHost?: string;
  } = {
    expectedRpIdHashHex,
    rpIdHashMatchesExpected: false,
  };

  try {
    const response = authenticationResponse.response as Record<string, unknown> | undefined;
    const clientDataJsonB64 =
      response && typeof response.clientDataJSON === 'string' ? response.clientDataJSON : undefined;
    const authenticatorDataB64 =
      response && typeof response.authenticatorData === 'string'
        ? response.authenticatorData
        : undefined;

    if (clientDataJsonB64) {
      const clientDataJson = Buffer.from(base64UrlToBytes(clientDataJsonB64)).toString('utf8');
      const clientData = JSON.parse(clientDataJson) as { origin?: string };
      inspection.clientDataOrigin = typeof clientData.origin === 'string' ? clientData.origin : undefined;
      if (inspection.clientDataOrigin) {
        try {
          inspection.clientDataOriginHost = new URL(inspection.clientDataOrigin).hostname.toLowerCase();
        } catch {
          inspection.clientDataOriginHost = inspection.clientDataOrigin.toLowerCase();
        }
      }
    }

    if (authenticatorDataB64) {
      const authenticatorData = base64UrlToBytes(authenticatorDataB64);
      if (authenticatorData.length >= 32) {
        inspection.authenticatorRpIdHashHex = bytesToHex(authenticatorData.slice(0, 32)).toLowerCase();
      }
    }
  } catch {
    // best-effort diagnostics only
  }

  inspection.rpIdHashMatchesExpected =
    inspection.authenticatorRpIdHashHex != null &&
    inspection.authenticatorRpIdHashHex.toLowerCase() === expectedRpIdHashHex;
  return inspection;
}

async function fetchJson<T>(url: string, init?: RequestInit): Promise<T> {
  const response = await fetch(url, init);
  const bodyText = await response.text();
  if (!response.ok) {
    throw new Error(`Request failed (${response.status}) ${url}: ${bodyText}`);
  }

  try {
    return JSON.parse(bodyText) as T;
  } catch {
    throw new Error(`Request returned non-JSON payload for ${url}: ${bodyText}`);
  }
}

async function fetchLatestBlockhashInfo(): Promise<{ blockhash: string }> {
  const data = await fetchJson<{ blockhash?: unknown }>(BLOCKHASH_URL);
  return { blockhash: requireString(data.blockhash, 'latest blockhash') };
}

function normalizeAuthenticationResponse(
  authenticationResponse: Record<string, unknown>,
): Record<string, unknown> {
  const response = (authenticationResponse.response ?? {}) as Record<string, unknown>;

  const rawIdInput = requireString(authenticationResponse.rawId, 'authentication response rawId');
  const normalizedRawId = canonicalizeCredentialId(rawIdInput);
  const normalizedId =
    typeof authenticationResponse.id === 'string' && authenticationResponse.id.trim().length > 0
      ? canonicalizeCredentialId(authenticationResponse.id)
      : normalizedRawId;

  const normalizeMaybeB64 = (field: unknown): unknown => {
    if (typeof field !== 'string' || field.trim().length === 0) return field;
    try {
      return canonicalizeBase64Like(field);
    } catch {
      return field;
    }
  };

  const normalizedResponse: Record<string, unknown> = {
    ...response,
    clientDataJSON: normalizeMaybeB64(response.clientDataJSON),
    authenticatorData: normalizeMaybeB64(response.authenticatorData),
    signature: normalizeMaybeB64(response.signature),
  };

  if (typeof response.userHandle === 'string') {
    normalizedResponse.userHandle = normalizeMaybeB64(response.userHandle);
  } else if (response.userHandle == null) {
    normalizedResponse.userHandle = null;
  }

  return {
    ...authenticationResponse,
    id: normalizedId,
    rawId: normalizedRawId,
    type:
      typeof authenticationResponse.type === 'string' && authenticationResponse.type.trim().length > 0
        ? authenticationResponse.type
        : 'public-key',
    clientExtensionResults:
      authenticationResponse.clientExtensionResults &&
      typeof authenticationResponse.clientExtensionResults === 'object'
        ? authenticationResponse.clientExtensionResults
        : {},
    response: normalizedResponse,
  };
}

async function getNativeWebAuthnAssertion(
  passkeyRpId: string,
  registrationRawId?: string,
): Promise<NativePasskeyAuthData> {
  const before = await fetchLatestBlockhashInfo();
  const challenge = toBase64Url(hexToBytes(before.blockhash));

  const authenticationOptions: Record<string, unknown> = {
    challenge,
    timeout: 60000,
    userVerification: 'required',
    rpId: passkeyRpId,
  };

  if (registrationRawId) {
    authenticationOptions.allowCredentials = [
      {
        id: registrationRawId,
        type: 'public-key',
      },
    ];
  }

  const authenticationResult = await litRustNativeGetPasskey(authenticationOptions);
  const rawAuthenticationResponse = JSON.parse(
    authenticationResult.authenticationResponseJson,
  ) as Record<string, unknown>;
  const authenticationResponse = normalizeAuthenticationResponse(rawAuthenticationResponse);

  const rawId = requireString(authenticationResponse.rawId, 'authentication response rawId');
  const authMethodId = deriveWebAuthnAuthMethodId(rawId);
  const inspection = inspectWebAuthnAssertion(authenticationResponse, passkeyRpId);

  if (!inspection.rpIdHashMatchesExpected) {
    throw new Error(
      `Native passkey assertion RP hash mismatch before Lit call. ` +
        `expected=${inspection.expectedRpIdHashHex} ` +
        `actual=${inspection.authenticatorRpIdHashHex || '(missing)'} ` +
        `origin=${inspection.clientDataOrigin || '(missing)'}`,
    );
  }

  return {
    authMethodType: AUTH_METHOD_TYPE_WEBAUTHN,
    authMethodId,
    accessToken: JSON.stringify(authenticationResponse),
  };
}

async function getRegistrationOptions(
  authServiceBaseUrl: string,
  username: string,
  expectedRpId: string,
  rpName: string,
): Promise<Record<string, unknown>> {
  const params = new URLSearchParams({
    username: requireString(username, 'username'),
  });
  const options = await fetchJson<Record<string, unknown>>(
    `${authServiceBaseUrl}/auth/webauthn/generate-registration-options?${params.toString()}`,
    {
      headers: buildAuthServiceHeaders(expectedRpId),
    },
  );

  const rp = (options.rp ?? {}) as Record<string, unknown>;
  const returnedRpId = typeof rp.id === 'string' ? rp.id.toLowerCase() : '';
  if (returnedRpId && returnedRpId !== expectedRpId) {
    throw new Error(
      `Auth service returned rp.id=${returnedRpId} but expected passkeyRpId=${expectedRpId}`,
    );
  }

  options.rp = {
    ...rp,
    id: expectedRpId,
    name: requireString(rpName, 'rpName'),
  };
  return options;
}

async function mintPkpWithAuthService(
  authServiceBaseUrl: string,
  passkeyRpId: string,
  registrationAuthMethodId: string,
  credentialPublicKey: string,
  scopes: string[],
): Promise<unknown> {
  const mintResponse = await fetch(`${authServiceBaseUrl}/pkp/mint`, {
    method: 'POST',
    headers: {
      'Content-Type': 'application/json',
      ...buildAuthServiceHeaders(passkeyRpId),
    },
    body: JSON.stringify({
      authMethodType: AUTH_METHOD_TYPE_WEBAUTHN,
      authMethodId: registrationAuthMethodId,
      pubkey: credentialPublicKey,
      scopes,
    }),
  });

  if (mintResponse.status !== 202) {
    const body = await mintResponse.text();
    throw new Error(`PKP mint failed: ${mintResponse.status} ${body}`);
  }

  const mintStart = (await mintResponse.json()) as { jobId?: unknown };
  const jobId = requireString(mintStart.jobId, 'pkp mint jobId');

  for (let attempt = 1; attempt <= MINT_STATUS_MAX_ATTEMPTS; attempt += 1) {
    await sleep(MINT_STATUS_POLL_DELAY_MS);
    const status = await fetchJson<{ state?: unknown; returnValue?: { data?: unknown } }>(
      `${authServiceBaseUrl}/status/${jobId}`,
      {
        headers: buildAuthServiceHeaders(passkeyRpId),
      },
    );
    const state = String(status.state || '').toLowerCase();

    if (state === 'completed' && status.returnValue?.data != null) {
      return status.returnValue.data;
    }

    if (state === 'failed' || state === 'error') {
      throw new Error(`PKP mint job failed: ${JSON.stringify(status)}`);
    }
  }

  throw new Error('PKP mint timed out');
}

export async function registerAndMintPKPWithNativePasskey(params: {
  network: string;
  rpcUrl: string;
  authServiceBaseUrl: string;
  expectedRpId: string;
  username?: string;
  rpName?: string;
  scopes?: string[];
}): Promise<NativePasskeyRegisterAndMintResult> {
  const authServiceBaseUrl = sanitizeBaseUrl(params.authServiceBaseUrl);
  const passkeyRpId = normalizeRpIdHost(params.expectedRpId);
  const username = params.username?.trim() ? String(params.username) : 'Heaven';
  const rpName = params.rpName?.trim() ? String(params.rpName) : 'Heaven';
  const scopes = params.scopes && params.scopes.length > 0 ? params.scopes : DEFAULT_SCOPES;

  const registrationOptions = await getRegistrationOptions(
    authServiceBaseUrl,
    username,
    passkeyRpId,
    rpName,
  );

  const registrationResult = await litRustNativeCreatePasskey(registrationOptions);
  const registrationResponse = JSON.parse(
    registrationResult.registrationResponseJson,
  ) as Record<string, unknown>;
  const registrationRawId = canonicalizeCredentialId(
    requireString(registrationResponse.rawId, 'registration response rawId'),
  );
  const registrationAuthMethodId = deriveWebAuthnAuthMethodId(registrationRawId);

  const pkpInfo = await mintPkpWithAuthService(
    authServiceBaseUrl,
    passkeyRpId,
    registrationAuthMethodId,
    requireString(registrationResult.credentialPublicKey, 'credentialPublicKey'),
    scopes,
  );

  // Warm the Lit node handshake before prompting the passkey so the access token stays fresh.
  await litRustTestConnect(params.network, requireString(params.rpcUrl, 'rpcUrl'));

  const authData = await getNativeWebAuthnAssertion(passkeyRpId, registrationRawId);

  return {
    pkpInfo,
    webAuthnPublicKey: registrationResult.credentialPublicKey,
    authData,
  };
}

export async function authenticateWithNativePasskey(params: {
  expectedRpId: string;
}): Promise<NativePasskeyAuthData> {
  const passkeyRpId = normalizeRpIdHost(params.expectedRpId);
  return getNativeWebAuthnAssertion(passkeyRpId);
}

File: apps/kotlin-app/app/src/main/java/com/pirate/app/lit/NativePasskeyAuth.kt
package com.pirate.app.lit

import android.app.Activity
import android.util.Base64
import androidx.credentials.CreateCredentialResponse
import androidx.credentials.CreatePublicKeyCredentialRequest
import androidx.credentials.CreatePublicKeyCredentialResponse
import androidx.credentials.CredentialManager
import androidx.credentials.CredentialManagerCallback
import androidx.credentials.GetCredentialRequest
import androidx.credentials.GetCredentialResponse
import androidx.credentials.GetPublicKeyCredentialOption
import androidx.credentials.PublicKeyCredential
import androidx.credentials.exceptions.CreateCredentialException
import androidx.credentials.exceptions.GetCredentialException
import com.upokecenter.cbor.CBORObject
import kotlinx.coroutines.Dispatchers
import kotlinx.coroutines.delay
import kotlinx.coroutines.withContext
import okhttp3.MediaType.Companion.toMediaType
import okhttp3.OkHttpClient
import okhttp3.Request
import okhttp3.RequestBody.Companion.toRequestBody
import org.bouncycastle.jcajce.provider.digest.Keccak
import org.json.JSONArray
import org.json.JSONObject
import java.io.ByteArrayInputStream
import java.security.MessageDigest
import kotlin.coroutines.resume
import kotlin.coroutines.resumeWithException
import kotlin.coroutines.suspendCoroutine

object NativePasskeyAuth {
  const val AUTH_METHOD_TYPE_WEBAUTHN = 3
  private const val BLOCKHASH_URL = "https://block-indexer.litgateway.com/get_most_recent_valid_block"
  private const val MINT_STATUS_MAX_ATTEMPTS = 20
  private const val MINT_STATUS_POLL_DELAY_MS = 3000L

  private val client = OkHttpClient()
  private val jsonMediaType = "application/json; charset=utf-8".toMediaType()

  data class AuthData(
    val authMethodType: Int,
    val authMethodId: String,
    val accessToken: String,
  )

  data class RegisterAndMintResult(
    val pkpInfo: JSONObject,
    val webAuthnPublicKey: String,
    val authData: AuthData,
  )

  suspend fun registerAndMintPkp(
    activity: Activity,
    authServiceBaseUrl: String,
    expectedRpId: String,
    username: String,
    rpName: String,
    scopes: List<String>,
    litNetwork: String,
    litRpcUrl: String,
  ): RegisterAndMintResult {
    val baseUrl = sanitizeBaseUrl(authServiceBaseUrl)
    val rpIdHost = normalizeRpIdHost(expectedRpId)

    val registrationOptions = getRegistrationOptions(
      authServiceBaseUrl = baseUrl,
      username = username,
      expectedRpId = rpIdHost,
      rpName = rpName,
    )

    val registration = createPasskey(activity, registrationOptions.toString())
    val registrationResponse = JSONObject(registration.registrationResponseJson)
    val registrationRawId = canonicalizeCredentialId(
      requireString(registrationResponse.optString("rawId", ""), "registration response rawId"),
    )
    val registrationAuthMethodId = deriveWebAuthnAuthMethodId(registrationRawId)

    val pkpInfo = mintPkpWithAuthService(
      authServiceBaseUrl = baseUrl,
      passkeyRpId = rpIdHost,
      registrationAuthMethodId = registrationAuthMethodId,
      credentialPublicKey = registration.credentialPublicKey,
      scopes = scopes,
    )

    // Warm Lit node handshake before prompting passkey for assertion.
    withContext(Dispatchers.IO) {
      LitRust.testConnectRaw(litNetwork, litRpcUrl)
    }

    val authData = getNativeWebAuthnAssertion(
      activity = activity,
      passkeyRpId = rpIdHost,
      registrationRawId = registrationRawId,
    )

    return RegisterAndMintResult(
      pkpInfo = pkpInfo,
      webAuthnPublicKey = registration.credentialPublicKey,
      authData = authData,
    )
  }

  suspend fun authenticateWithPasskey(
    activity: Activity,
    expectedRpId: String,
  ): AuthData {
    val rpIdHost = normalizeRpIdHost(expectedRpId)
    return getNativeWebAuthnAssertion(activity = activity, passkeyRpId = rpIdHost, registrationRawId = null)
  }

  private suspend fun getNativeWebAuthnAssertion(
    activity: Activity,
    passkeyRpId: String,
    registrationRawId: String?,
  ): AuthData {
    val before = fetchLatestBlockhashInfo()
    val challenge = toBase64Url(hexToBytes(before.getString("blockhash")))

    val authenticationOptions = JSONObject()
      .put("challenge", challenge)
      .put("timeout", 60000)
      .put("userVerification", "required")
      .put("rpId", passkeyRpId)

    if (registrationRawId != null) {
      val allowCredentials = JSONArray()
        .put(JSONObject().put("id", registrationRawId).put("type", "public-key"))
      authenticationOptions.put("allowCredentials", allowCredentials)
    }

    val authentication = getPasskey(activity, authenticationOptions.toString())
    val rawAuthenticationResponse = JSONObject(authentication.authenticationResponseJson)
    val normalized = normalizeAuthenticationResponse(rawAuthenticationResponse)

    val rawId = requireString(normalized.optString("rawId", ""), "authentication response rawId")
    val authMethodId = deriveWebAuthnAuthMethodId(rawId)

    val inspection = inspectWebAuthnAssertion(normalized, passkeyRpId)
    if (!inspection.rpIdHashMatchesExpected) {
      throw IllegalStateException(
        "Native passkey assertion RP hash mismatch before Lit call. " +
          "expected=${inspection.expectedRpIdHashHex} actual=${inspection.authenticatorRpIdHashHex ?: "(missing)"} " +
          "origin=${inspection.clientDataOrigin ?: "(missing)"}",
      )
    }

    return AuthData(
      authMethodType = AUTH_METHOD_TYPE_WEBAUTHN,
      authMethodId = authMethodId,
      accessToken = normalized.toString(),
    )
  }

  private fun sanitizeBaseUrl(value: String): String {
    val trimmed = requireString(value, "authServiceBaseUrl").trim()
    return trimmed.replace(Regex("/+$"), "")
  }

  private fun requireString(value: String?, fieldName: String): String {
    val parsed = value?.trim().orEmpty()
    if (parsed.isEmpty()) throw IllegalArgumentException("$fieldName cannot be empty")
    return parsed
  }

  private fun parseHostname(value: String, fieldName: String): String {
    val trimmed = requireString(value, fieldName)
    val withScheme = if (Regex("^[a-z][a-z0-9+.-]*://", RegexOption.IGNORE_CASE).containsMatchIn(trimmed)) {
      trimmed
    } else {
      "https://$trimmed"
    }
    val host = try {
      java.net.URL(withScheme).host
    } catch (_: Throwable) {
      throw IllegalArgumentException("Invalid $fieldName: $value")
    }
    val normalized = host.trim().lowercase()
    if (normalized.isEmpty()) throw IllegalArgumentException("Invalid $fieldName: $value")
    return normalized
  }

  private fun normalizeRpIdHost(value: String): String = parseHostname(value, "passkeyRpId")

  private fun buildAuthServiceHeaders(passkeyRpId: String): Map<String, String> {
    val origin = "https://$passkeyRpId"
    return mapOf(
      "Origin" to origin,
      "Referer" to "$origin/",
    )
  }

  private suspend fun fetchJson(url: String, headers: Map<String, String> = emptyMap()): JSONObject {
    return withContext(Dispatchers.IO) {
      val req = Request.Builder().url(url).get().apply {
        headers.forEach { (k, v) -> addHeader(k, v) }
      }.build()
      client.newCall(req).execute().use { resp ->
        val body = resp.body?.string().orEmpty()
        if (!resp.isSuccessful) {
          throw IllegalStateException("Request failed (${resp.code}) $url: $body")
        }
        try {
          JSONObject(body)
        } catch (_: Throwable) {
          throw IllegalStateException("Request returned non-JSON payload for $url: $body")
        }
      }
    }
  }

  private suspend fun fetchLatestBlockhashInfo(): JSONObject {
    val data = fetchJson(BLOCKHASH_URL)
    val blockhash = data.optString("blockhash", "").trim()
    if (blockhash.isEmpty()) throw IllegalStateException("latest blockhash cannot be empty")
    return JSONObject().put("blockhash", blockhash)
  }

  private suspend fun getRegistrationOptions(
    authServiceBaseUrl: String,
    username: String,
    expectedRpId: String,
    rpName: String,
  ): JSONObject {
    val url = "${authServiceBaseUrl}/auth/webauthn/generate-registration-options?username=${java.net.URLEncoder.encode(username, "UTF-8")}"
    val options = fetchJson(url, headers = buildAuthServiceHeaders(expectedRpId))

    val rp = options.optJSONObject("rp") ?: JSONObject()
    val returnedRpId = rp.optString("id", "").trim().lowercase()
    if (returnedRpId.isNotEmpty() && returnedRpId != expectedRpId) {
      throw IllegalStateException("Auth service returned rp.id=$returnedRpId but expected passkeyRpId=$expectedRpId")
    }

    rp.put("id", expectedRpId)
    rp.put("name", requireString(rpName, "rpName"))
    options.put("rp", rp)
    return options
  }

  private suspend fun mintPkpWithAuthService(
    authServiceBaseUrl: String,
    passkeyRpId: String,
    registrationAuthMethodId: String,
    credentialPublicKey: String,
    scopes: List<String>,
  ): JSONObject {
    val bodyJson = JSONObject()
      .put("authMethodType", AUTH_METHOD_TYPE_WEBAUTHN)
      .put("authMethodId", registrationAuthMethodId)
      .put("pubkey", requireString(credentialPublicKey, "credentialPublicKey"))
      .put("scopes", JSONArray(scopes))

    val req = Request.Builder()
      .url("${authServiceBaseUrl}/pkp/mint")
      .post(bodyJson.toString().toRequestBody(jsonMediaType))
      .apply {
        addHeader("Content-Type", "application/json")
        buildAuthServiceHeaders(passkeyRpId).forEach { (k, v) -> addHeader(k, v) }
      }
      .build()

    val jobId = withContext(Dispatchers.IO) {
      client.newCall(req).execute().use { resp ->
        val body = resp.body?.string().orEmpty()
        if (resp.code != 202) {
          throw IllegalStateException("PKP mint failed: ${resp.code} $body")
        }
        val parsed = JSONObject(body)
        requireString(parsed.optString("jobId", ""), "pkp mint jobId")
      }
    }

    for (attempt in 1..MINT_STATUS_MAX_ATTEMPTS) {
      delay(MINT_STATUS_POLL_DELAY_MS)
      val status = fetchJson("${authServiceBaseUrl}/status/${jobId}", headers = buildAuthServiceHeaders(passkeyRpId))
      val state = status.optString("state", "").trim().lowercase()

      if (state == "completed") {
        val returnValue = status.optJSONObject("returnValue")
        val data = returnValue?.opt("data")
        if (data is JSONObject) return data
        if (data != null) return JSONObject(data.toString())
      }

      if (state == "failed" || state == "error") {
        throw IllegalStateException("PKP mint job failed: ${status}")
      }
    }

    throw IllegalStateException("PKP mint timed out")
  }

  private fun hexToBytes(value: String): ByteArray {
    val normalized = value.removePrefix("0x").removePrefix("0X")
    if (!Regex("^[0-9a-fA-F]+$").matches(normalized) || normalized.length % 2 != 0) {
      throw IllegalArgumentException("Invalid hex value: $value")
    }
    val out = ByteArray(normalized.length / 2)
    var i = 0
    while (i < normalized.length) {
      out[i / 2] = normalized.substring(i, i + 2).toInt(16).toByte()
      i += 2
    }
    return out
  }

  private fun toBase64Url(value: ByteArray): String {
    val base64 = Base64.encodeToString(value, Base64.NO_WRAP)
    return base64.replace('+', '-').replace('/', '_').trimEnd('=')
  }

  private fun base64UrlToBytes(value: String): ByteArray {
    val normalized = value.replace('-', '+').replace('_', '/')
    val padded = normalized + "=".repeat((4 - (normalized.length % 4)) % 4)
    return Base64.decode(padded, Base64.DEFAULT)
  }

  private fun canonicalizeBase64Like(value: String): String {
    return toBase64Url(base64UrlToBytes(value))
  }

  private fun canonicalizeCredentialId(value: String): String = canonicalizeBase64Like(value)

  private fun bytesToHex(value: ByteArray): String {
    val sb = StringBuilder(value.size * 2)
    for (b in value) sb.append(String.format("%02x", b.toInt() and 0xFF))
    return sb.toString()
  }

  private fun deriveWebAuthnAuthMethodId(rawId: String): String {
    val input = "$rawId:lit".toByteArray(Charsets.UTF_8)
    val digest = Keccak.Digest256().digest(input)
    return "0x${bytesToHex(digest)}"
  }

  private data class AssertionInspection(
    val expectedRpIdHashHex: String,
    val authenticatorRpIdHashHex: String?,
    val rpIdHashMatchesExpected: Boolean,
    val clientDataOrigin: String?,
  )

  private fun inspectWebAuthnAssertion(authenticationResponse: JSONObject, expectedRpId: String): AssertionInspection {
    val expectedHashBytes = MessageDigest.getInstance("SHA-256")
      .digest(expectedRpId.toByteArray(Charsets.UTF_8))
    val expectedHashHex = bytesToHex(expectedHashBytes).lowercase()

    var clientDataOrigin: String? = null
    var authenticatorRpIdHashHex: String? = null

    try {
      val response = authenticationResponse.optJSONObject("response") ?: JSONObject()
      val clientDataJsonB64 = response.optString("clientDataJSON", "").takeIf { it.isNotBlank() }
      val authenticatorDataB64 = response.optString("authenticatorData", "").takeIf { it.isNotBlank() }

      if (clientDataJsonB64 != null) {
        val clientDataJson = String(base64UrlToBytes(clientDataJsonB64), Charsets.UTF_8)
        val clientData = JSONObject(clientDataJson)
        clientDataOrigin = clientData.optString("origin", "").takeIf { it.isNotBlank() }
      }

      if (authenticatorDataB64 != null) {
        val authenticatorData = base64UrlToBytes(authenticatorDataB64)
        if (authenticatorData.size >= 32) {
          authenticatorRpIdHashHex = bytesToHex(authenticatorData.copyOfRange(0, 32)).lowercase()
        }
      }
    } catch (_: Throwable) {
      // best-effort only
    }

    val matches = authenticatorRpIdHashHex != null && authenticatorRpIdHashHex == expectedHashHex
    return AssertionInspection(
      expectedRpIdHashHex = expectedHashHex,
      authenticatorRpIdHashHex = authenticatorRpIdHashHex,
      rpIdHashMatchesExpected = matches,
      clientDataOrigin = clientDataOrigin,
    )
  }

  private fun normalizeAuthenticationResponse(authenticationResponse: JSONObject): JSONObject {
    val response = authenticationResponse.optJSONObject("response") ?: JSONObject()

    val rawIdInput = requireString(authenticationResponse.optString("rawId", ""), "authentication response rawId")
    val normalizedRawId = canonicalizeCredentialId(rawIdInput)

    val normalizedId = authenticationResponse.optString("id", "").trim().takeIf { it.isNotEmpty() }?.let {
      canonicalizeCredentialId(it)
    } ?: normalizedRawId

    fun normalizeMaybeB64(field: Any?): Any? {
      val asString = field as? String ?: return field
      if (asString.trim().isEmpty()) return field
      return try { canonicalizeBase64Like(asString) } catch (_: Throwable) { field }
    }

    val normalizedResponse = JSONObject(response.toString())
    normalizedResponse.put("clientDataJSON", normalizeMaybeB64(response.opt("clientDataJSON")))
    normalizedResponse.put("authenticatorData", normalizeMaybeB64(response.opt("authenticatorData")))
    normalizedResponse.put("signature", normalizeMaybeB64(response.opt("signature")))
    if (response.has("userHandle")) {
      normalizedResponse.put("userHandle", normalizeMaybeB64(response.opt("userHandle")))
    } else {
      normalizedResponse.put("userHandle", JSONObject.NULL)
    }

    val out = JSONObject(authenticationResponse.toString())
    out.put("id", normalizedId)
    out.put("rawId", normalizedRawId)
    out.put("type", authenticationResponse.optString("type", "public-key").ifBlank { "public-key" })
    if (authenticationResponse.opt("clientExtensionResults") !is JSONObject) {
      out.put("clientExtensionResults", JSONObject())
    }
    out.put("response", normalizedResponse)
    return out
  }

  data class CreatePasskeyResult(
    val registrationResponseJson: String,
    val credentialPublicKey: String,
  )

  data class GetPasskeyResult(
    val authenticationResponseJson: String,
  )

  private suspend fun createPasskey(activity: Activity, optionsJson: String): CreatePasskeyResult {
    val manager = CredentialManager.create(activity)
    val request = CreatePublicKeyCredentialRequest(optionsJson)

    return suspendCoroutine { cont ->
      manager.createCredentialAsync(
        activity,
        request,
        null,
        activity.mainExecutor,
        object : CredentialManagerCallback<CreateCredentialResponse, CreateCredentialException> {
          override fun onResult(result: CreateCredentialResponse) {
            try {
              if (result !is CreatePublicKeyCredentialResponse) {
                cont.resumeWithException(IllegalStateException("Unexpected credential response type: ${result::class.java.simpleName}"))
                return
              }
              val registrationResponseJson = result.registrationResponseJson
              val credentialPublicKey = extractCredentialPublicKeyHexFromRegistrationResponse(registrationResponseJson)
              cont.resume(CreatePasskeyResult(registrationResponseJson, credentialPublicKey))
            } catch (t: Throwable) {
              cont.resumeWithException(t)
            }
          }

          override fun onError(e: CreateCredentialException) {
            cont.resumeWithException(e)
          }
        },
      )
    }
  }

  private suspend fun getPasskey(activity: Activity, optionsJson: String): GetPasskeyResult {
    val manager = CredentialManager.create(activity)
    val request = GetCredentialRequest(listOf(GetPublicKeyCredentialOption(optionsJson)))

    return suspendCoroutine { cont ->
      manager.getCredentialAsync(
        activity,
        request,
        null,
        activity.mainExecutor,
        object : CredentialManagerCallback<GetCredentialResponse, GetCredentialException> {
          override fun onResult(result: GetCredentialResponse) {
            val credential = result.credential
            if (credential !is PublicKeyCredential) {
              cont.resumeWithException(IllegalStateException("Unexpected credential type: ${credential::class.java.simpleName}"))
              return
            }
            cont.resume(GetPasskeyResult(authenticationResponseJson = credential.authenticationResponseJson))
          }

          override fun onError(e: GetCredentialException) {
            cont.resumeWithException(e)
          }
        },
      )
    }
  }

  private fun extractCredentialPublicKeyHexFromRegistrationResponse(registrationResponseJson: String): String {
    val registrationJson = JSONObject(registrationResponseJson)
    val response = registrationJson.optJSONObject("response")
      ?: throw IllegalArgumentException("registration response missing `response` object")
    val attestationObject = response.optString("attestationObject", "")
    if (attestationObject.isBlank()) {
      throw IllegalArgumentException("registration response missing `response.attestationObject`")
    }

    val attestationBytes = base64UrlToBytes(attestationObject)
    val attestation = CBORObject.DecodeFromBytes(attestationBytes)
    val authDataObject = attestation.get("authData")
      ?: throw IllegalArgumentException("attestation object missing `authData` field")
    val authData = authDataObject.GetByteString()
    return extractCredentialPublicKeyHexFromAuthData(authData)
  }

  private fun extractCredentialPublicKeyHexFromAuthData(authData: ByteArray): String {
    // authData format: rpIdHash(32) | flags(1) | signCount(4) | attestedCredentialData...
    if (authData.size < 55) {
      throw IllegalArgumentException("authData too short to contain attested credential data")
    }

    val flags = authData[32].toInt() and 0xFF
    val hasAttestedCredentialData = (flags and 0x40) != 0
    if (!hasAttestedCredentialData) {
      throw IllegalArgumentException("attested credential data flag not set in authData")
    }

    var offset = 37 // 32 + 1 + 4
    offset += 16 // aaguid
    if (authData.size < offset + 2) {
      throw IllegalArgumentException("authData missing credentialId length")
    }

    val credentialIdLength =
      ((authData[offset].toInt() and 0xFF) shl 8) or (authData[offset + 1].toInt() and 0xFF)
    offset += 2
    if (authData.size < offset + credentialIdLength) {
      throw IllegalArgumentException("authData truncated credentialId bytes")
    }
    offset += credentialIdLength
    if (authData.size <= offset) {
      throw IllegalArgumentException("authData missing CBOR-encoded credential public key")
    }

    val remaining = authData.copyOfRange(offset, authData.size)
    val stream = ByteArrayInputStream(remaining)
    if (CBORObject.Read(stream) == null) {
      throw IllegalArgumentException("failed to parse CBOR credential public key")
    }
    val consumed = remaining.size - stream.available()
    if (consumed <= 0) {
      throw IllegalArgumentException("credential public key CBOR length could not be determined")
    }

    val credentialPublicKeyBytes = remaining.copyOfRange(0, consumed)
    return "0x" + credentialPublicKeyBytes.joinToString("") { byte ->
      "%02x".format(byte.toInt() and 0xFF)
    }
  }
}

